### PR TITLE
Fix conv x-backward padding_out bug

### LINKED ISSUE #4799: Autodiff backward panics on Conv1d(asymmetric padding) → GroupNorm → Conv1d for non-trivial input lengths
## Summary

`loss.backward()` panics when the forward graph is `Conv1d(asymmetric padding) → GroupNorm → Conv1d` and the input is long enough that the first conv's output has more than roughly 160 frames. The panic surfaces in both `NdArray` and `Flex` backends with different messages, so the root cause lives in burn's shared asymmetric-padding handling for `Conv1d`, not in any single backend.

Both conv layers use `PaddingConfig1d::Explicit(left, right)` with `right != left`. Forward produces correct output shapes on both backends; only backward fails.

## Minimal reproducer

```rust
use burn::nn::conv::Conv1dConfig;
use burn::nn::{GroupNormConfig, PaddingConfig1d};
use burn::prelude::*;
use burn::tensor::Distribution;
use burn::tensor::backend::AutodiffBackend;

fn run<B: AutodiffBackend>(device: &B::Device, input_length: usize) {
    let c1 = Conv1dConfig::new(1, 32, 10)
        .with_stride(10)
        .with_padding(PaddingConfig1d::Explicit(9, 0))
        .init::<B>(device);
    let gn = GroupNormConfig::new(8, 32).init::<B>(device);
    let c2 = Conv1dConfig::new(32, 64, 4)
        .with_stride(4)
        .with_padding(PaddingConfig1d::Explicit(3, 0))
        .init::<B>(device);

    let x = Tensor::<B, 3>::random([1, 1, input_length], Distribution::Default, device);
    let h = c1.forward(x);
    let h = gn.forward(h);
    let h = c2.forward(h);
    let _ = h.sum().backward();  // panics
}

#[test]
fn ndarray_panics() {
    use burn::backend::{Autodiff, NdArray};
    run::<Autodiff<NdArray>>(&Default::default(), 16_000);
}

#[test]
fn flex_panics() {
    use burn::backend::Autodiff;
    use burn_flex::Flex;
    run::<Autodiff<Flex>>(&Default::default(), 16_000);
}
```

## Expected behaviour

`loss.backward()` returns without panic and produces gradients with shapes matching the parameters (as it does on the equivalent `Conv1d + Conv1d` graph without the GroupNorm in between, and as it does at small input lengths).

### NdArray

```
panicked at ndarray-0.17.2/src/impl_methods.rs:591:22:
Slice end 1603 is past end of axis of length 1601
```

Stack trace points into `burn_autodiff::ops::backward::binary` feeding `burn_ndarray::ops::base::NdArrayOps::slice`, called from the autodiff step for `float_slice_assign`.

### Flex

```
panicked at burn-autodiff-0.21.0-pre.3/src/ops/base.rs:320:17:
Invalid broadcast shapes:
  Next grad shape Shape { dims: [1, 32, 1600] },
  Previous grad shape Shape { dims: [1, 32, 1598] }.
Expected the shape of the next grad to be 1.
```

## Input-length sweep (Flex, same two-conv + GroupNorm graph)

| `input_length` | `c1` output frames | backward |
| --: | --: | :-- |
| 160 | 16 | OK |
| 1 600 | 160 | OK |
| 16 000 | 1 600 | **panic** (1600 vs 1598) |
| 32 000 | 3 200 | **panic** (3200 vs 3198) |
| 64 000 | 6 400 | **panic** (16000 vs 15998) |
| 160 000 | 16 000 | **panic** (32000 vs 31998) |
| 320 000 | 32 000 | **panic** (33856 vs 33854) |
| 338 560 | 33 856 | **panic** (33856 vs 33854) |

The grad-shape mismatch is always exactly 2 frames, which matches `c2`'s `left_pad = 3` minus a likely off-by-one somewhere in the asymmetric-pad backward.

## Narrowing

- `c1 + c2` (no GroupNorm): **OK** on both backends.
- `c1 + GroupNorm` (no second conv): **OK** on both backends.
- `c1 + GroupNorm + c2` (above): **panics**.
- Same graph with symmetric padding (`PaddingConfig1d::Same` everywhere): **OK** (no asymmetric-pad path).

The critical ingredients are **asymmetric padding** on `c1` / `c2`, a non-identity op between them that materializes a new tensor in the autograd graph, and an input long enough to produce more than roughly 160 frames at the `c1` output. Removing any of the three makes backward succeed.

## Likely location

Wrapper in `burn_tensor::tensor::module::conv1d` (around burn-tensor 0.21.0-pre.3 line 47):

```rust
if let Some(padding_end) = padded_options.padding_end {
    let left = padded_options.options.padding[0];
    let right = padding_end[0];
    let padded = x.pad((left, right, 0, 0), PadMode::Constant(0.0));
    let zero_options = ConvOptions::new(..., [0], ...);
    // backend conv1d with padding=0 on the manually-padded tensor
}
```

The asymmetric-pad case routes through `x.pad(...)` + backend `conv1d(padding=0)`. The backward traversal through this pair appears to drop or miscompute the extra `left + right` padding slices, producing a gradient tensor whose `T` dim is `left + right` shorter than the upstream gradient expects.

## Versions

- `burn = "0.21.0-pre.3"`
- `burn-flex = "0.21.0-pre.3"`
- Rust 1.94
- macOS 15.4 (darwin 25.4)

## Workaround

None that preserves asymmetric padding. `PaddingConfig1d::Same` works but is symmetric.

### PR DESCRIPTION
**Pull request by antimora**
## Summary

Fixes #4799.

`calculate_padding_out` in `burn-backend::ops::modules::conv` feeds the transpose conv used by `conv{1,2,3}d_x_backward`. Its ceil/floor formula only accounted for 0- or 1-element drops, so when a forward conv dropped 2+ tail inputs (e.g. `kernel=4, stride=4, L_in=1603` drops 3), the backward transpose produced a grad tensor shorter than the original input and downstream pad/slice_assign ops panicked.

Replaced the formula with a direct inversion of the transpose conv output formula, which gives the exact padding regardless of how many elements the forward drops. Same helper is shared across 1D/2D/3D.

The bug was latent for many users because the first conv's input is usually a leaf that doesn't require grad, so `conv1d_x_backward` is skipped. The reported reproducer forces the second conv's x-backward to actually run by placing a GroupNorm between two asymmetric-padded convs.

## Test plan

- [x] `cargo test -p burn-backend-tests --test autodiff conv` - all 94 conv autodiff tests pass
- [x] `cargo test -p burn-backend-tests --test tensor conv` - all 45 conv tensor tests pass
- [x] Added `test_conv1d_backward_shape_with_remainder` - targeted test of the dropped-inputs path
- [x] Added `test_conv1d_asymmetric_padding_backward_shape` - end-to-end via `PaddedConvOptions::asymmetric` mirroring the issue scenario

### CODE REVIEW

**Review by laggui — DISMISSED** [review]
/edit: whoops commented on the wrong PR :smile:

**Review by laggui — COMMENTED** [review]
Good catch!

Looks like the same function exists in cubecl, we need to mirror the fix there

https://github.com/tracel-ai/burn/blob/main/crates/burn-cubecl/src/kernel/conv/backward_data/fallback.rs#L92

File: crates/burn-backend/src/backend/ops/modules/conv.rs [hunk 1]
```diff
@@ -1296,10 +1296,15 @@ fn calculate_padding_out(
...
-    i64::max(0, out as i64 - size_out as i64) as usize
+    // Invert the transpose conv output formula to recover the exact number of
+    // input elements that a forward conv would drop for this (size_in, size_out).
+    //
+    // Forward: size_out = floor((size_in + 2*padding - dilated_kernel) / stride) + 1
+    // Transpose: trans_out = (size_out - 1)*stride + dilated_kernel + padding_out - 2*padding
+    // Setting trans_out == size_in and solving for padding_out:
+    let dilated_kernel = dilation * (kernel_size - 1) + 1;
+    let base = (size_out as i64 - 1) * stride as i64 + dilated_kernel as i64 - 2 * padding as i64;
+    i64::max(0, size_in as i64 - base) as usize
```
[the remark this answers is not in the document]
Comment (antimora) [reply crates/burn-backend/src/backend/ops/modules/conv.rs:1311]: Good catch, fixed in 465e227c6 - mirrored the same inversion-of-transpose-output-size formula in `crates/burn-cubecl/src/kernel/conv/backward_data/fallback.rs`. Update: went further in cde6c3ef5 and deduplicated the two copies, exposing `calculate_padding_out` as `pub` in burn-backend and importing it in the CubeCL fallback.

File: crates/burn-backend-tests/tests/autodiff/conv1d.rs [hunk 2]
```diff
@@ -200,6 +204,71 @@ fn test_conv1d_groups() {
...
+    assert_eq!(x_grad.dims(), [1, 1, length]);
+    assert_eq!(weight_grad.dims(), [1, 1, 4]);
+}
+
+/// End-to-end regression mirroring the scenario from
+/// https://github.com/tracel-ai/burn/issues/4799: two conv1ds with asymmetric
+/// padding separated by an op that materializes a new tensor. The asymmetric
+/// path goes `x.pad(...) -> B::conv1d(padding=0)`, so a miscomputed transpose
+/// padding in the backward fails shape checks when the second conv drops
+/// multiple tail inputs.
```
[the remark this answers is not in the document]
Comment (antimora) [reply crates/burn-backend-tests/tests/autodiff/conv1d.rs]: Right, the doc was leftover from the issue description. Tightened in 465e227c6 to describe what the test actually does: a reduced single-conv1d case exercising the asymmetric-padding backward path.

**Comment by antimora** [comment]
@laggui Mirrored the fix in the CubeCL fallback in 465e227c6 - same inversion-of-transpose-output-size formula. Update: deduplicated in cde6c3ef5 - exposed the helper in burn-backend and made the CubeCL fallback import it, since burn-cubecl already depends on burn-backend.

**Comment by antimora** [comment]
Went ahead and deduplicated in cde6c3ef5 - exposed `calculate_padding_out` as `pub` in burn-backend and made the CubeCL fallback import it instead of carrying its own copy. burn-cubecl already depended on burn-backend, so it's a net negative on line count.

**Review by laggui — APPROVED** [review]
LGTM!
